onnxruntime/core/providers/cpu/llm/attention.cc

```diff
@@ -0,0 +1,749 @@
+
+    y_shape = {static_cast<int64_t>(parameters.batch_size),
+               static_cast<int64_t>(parameters.q_sequence_length),
+               static_cast<int64_t>(parameters.q_num_heads * parameters.v_head_size)};
+  }
+  ORT_ENFORCE(attn_mask == nullptr || attn_mask->Shape()[attn_mask->Shape().NumDimensions() - 1] == parameters.past_sequence_length + parameters.kv_sequence_length, "inconsistent total_sequence_length");
+  ORT_ENFORCE(past_key == nullptr || past_key->Shape()[1] == parameters.kv_num_heads, "inconsistent kv_num_heads");
+  ORT_ENFORCE(past_value == nullptr || past_value->Shape()[1] == parameters.kv_num_heads, "inconsistent kv_num_heads");
+  ORT_ENFORCE(past_key == nullptr || past_key->Shape()[3] == parameters.head_size, "inconsistent head_size");
+  ORT_ENFORCE(past_value == nullptr || past_value->Shape()[3] == parameters.v_head_size, "inconsistent v_head_size");
+  ORT_ENFORCE(past_key == nullptr || past_key->Shape()[2] == parameters.past_sequence_length, "inconsistent past_sequence_length");
+  ORT_ENFORCE(past_value == nullptr || past_value->Shape()[2] == parameters.past_sequence_length, "inconsistent past_sequence_length");
+
+  parameters.scale = std::isnan(scale_) ? static_cast<float>(1.0 / sqrt(parameters.head_size)) : scale_;
+  ORT_ENFORCE(parameters.getAttentionType() != AttentionType::kInvalid, "Invalid attention type. q_num_heads must be equal to kv_num_heads or a multiple of it.");
+
+  std::vector<int64_t> present_key_shape = {static_cast<int64_t>(parameters.batch_size),
+                                            static_cast<int64_t>(parameters.kv_num_heads),
+                                            static_cast<int64_t>(parameters.past_sequence_length + parameters.kv_sequence_length),
+                                            static_cast<int64_t>(parameters.head_size)};
+  std::vector<int64_t> present_value_shape = {static_cast<int64_t>(parameters.batch_size),
+                                              static_cast<int64_t>(parameters.kv_num_heads),
+                                              static_cast<int64_t>(parameters.past_sequence_length + parameters.kv_sequence_length),
+                                              static_cast<int64_t>(parameters.v_head_size)};
+  std::vector<int64_t> output_qk_shape = {static_cast<int64_t>(parameters.batch_size),
+                                          static_cast<int64_t>(parameters.q_num_heads),
+                                          static_cast<int64_t>(parameters.q_sequence_length),
+                                          static_cast<int64_t>(parameters.past_sequence_length + parameters.kv_sequence_length)};
+
+  Tensor* Y = context->Output(0, y_shape);
+  Tensor* present_key = context->Output(1, present_key_shape);
+  Tensor* present_value = context->Output(2, present_value_shape);
+  Tensor* output_qk = context->Output(3, output_qk_shape);
+
+  return this->ApplyAttention(context,
```

Comment:
In general, I think we should wrap the input check above into CheckInputs function (following ORT convention), because we can really reuse those lines when we implement CUDA Attention.